[TACHYON-599] Provide narrower views of BlockMetadataManager

ooq opened the pull request:
https://tachyon.atlassian.net/browse/TACHYON-599

apc999 commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,137 @@
+   *
+   * @return the InetSocketAddress of the master
+   */
+  private InetSocketAddress getMasterAddress() {
```
Please use `tachyon.worker.block.BlockWorkerUtils#getMasterAddress`

apc999 commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,137 @@
+            ThreadFactoryUtils.build("worker-client-pinlist-%d", true));
+    mMasterClient =
+        new MasterClient(getMasterAddress(), mMasterClientExecutorService, mTachyonConf);
+    mHeartbeatIntervalMs =
```
`mHeartbeatIntervalMs` -> `mFetchIntervalMs` ?

apc999 commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,137 @@
+  private void resetMasterClient() {
+    mMasterClient.close();
+    mMasterClient =
+        new MasterClient(getMasterAddress(), mMasterClientExecutorService, mTachyonConf);
```
again `getMasterAddress` -> `BlockWorkerUtils.getMasterAddress`

apc999 commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -74,6 +74,12 @@
   private final Evictor mEvictor;
   private List<BlockStoreEventListener> mBlockStoreEventListeners =
       new ArrayList<BlockStoreEventListener>();
+  /** A list of pinned blocks fetched from the master */
```
list -> set 
blocks -> inodes

apc999 commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -74,6 +74,12 @@
       new ArrayList<BlockStoreEventListener>();
+  /** A list of pinned blocks fetched from the master */
+  private final Set<Integer> mPinnedInodes = new HashSet<Integer>();
+  /** A list of blocks that are currently being locked */
```
list -> set

apc999 commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -413,6 +423,19 @@ private void removeBlockNoLock(long userId, long blockId) throws IOException {
     mMetaManager.removeBlockMeta(blockMeta);
   }
 
+  private void updateLockedBlocks() {
```
I don't think you need this method and `mLockedBlocks`, and you can simply call `mLockManager.getLockedBlocks` inside `getUpdatedView`

ooq replied:
Maybe the code is easier to understand this way?

ooq replied:
adopted your suggestions. thanks.

apc999 commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -74,6 +74,12 @@
+  /** A list of blocks that are currently being locked */
+  private final Set<Long> mLockedBlocks = new HashSet<Long>();
+  /** A narrowed view provided to evictors and allocators */
+  private BlockMetadataManagerView mMetadataView;
```
Since in this version we don't keep a long-living `BlockMetadataManagerView`, it is unnecessary to keep `mMetadataView`

ooq replied:
Oh, saw your comment below first. Yeah, we can do that.

apc999 commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -413,6 +423,19 @@ private void removeBlockNoLock(long userId, long blockId) throws IOException {
+  private BlockMetadataManagerView getUpdatedView() throws IOException {
+    // TODO: update the view object instead of creating new one every time
+    updateLockedBlocks();
+    mMetadataView = new BlockMetadataManagerView(mMetaManager, mPinnedInodes, mLockedBlocks);
```
why not simply return `new BlockMetadataManagerView(...)`

ooq replied:
The thing is we don't know whether we will make the view long-living: if the view is only short living, then we don't even need mMetadataView; if it is long living, others might access the view without doing the update.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockDataManager.java:
```diff
@@ -424,4 +425,14 @@ public void userHeartbeat(long userId, List<Long> metrics) {
+   * Set the pinlist for the underlying blockstore.
+   * Typically called by PinListSyncer.
+   *
+   * @param list
```
Please fill in the javadoc.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockLockManager.java:
```diff
@@ -200,6 +201,17 @@ public void cleanupUser(long userId) {
   }
 
   /**
+   * Get a set of currently locked blocks.
```
Please complete the javadoc with `@return`

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  /**
+   * Return a filtered list of block metadata,
```
better format this javadoc

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,102 @@
+   * This API is here so that to be compatible with Allocator implementation.
+   * Ideally it should be removed.
+   */
+  public StorageDir getDirForCreatingBlock() {
```
Instead of returning handler to underlying `mDir`, make a method to return `TempBlockMeta` ?

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockDataManager.java:
```diff
@@ -424,4 +425,14 @@ public void userHeartbeat(long userId, List<Long> metrics) {
+   *
+   * @param list
+   */
+  public void setPinList(Set<Integer> list) {
```
* This javadoc is not clear what are the integers in the list.
* describe parameter `list` in javadoc

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+   * Now we always creating a new view before freespace.
+   * TODO: incrementally update the view
+   *
+   * @param manager
```
update javadoc

ooq commented:
@apc999 @calvinjia addressed the comments.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+   * @param manager
+   * @param pinnedBlocks
+   * @param lockedBlocks
+   * @throws IOException
```
Please complete the javadoc.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockLockManager.java:
```diff
@@ -200,6 +201,19 @@ public void cleanupUser(long userId) {
+   */
+  public Set<Long> getLockedBlocks() {
+    Set<Long> set = new HashSet<Long>();
+    for (LockRecord lockRecord : mLockIdToRecordMap.values()) {
```
Do you need to put this in the `mSharedMapsLock`?

apc999 replied:
+1

ooq replied:
Good point, adding this lock so the method is independent of how people use it.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+      Set<Long> lockedBlocks) throws IOException {
+    mMetadataManager = Preconditions.checkNotNull(manager);
+    mPinnedInodes.addAll(Preconditions.checkNotNull(pinnedInodes));
+    mLockedBlocks.addAll(Preconditions.checkNotNull(lockedBlocks));
```
Instead of initializing the sets in the declaration, would it be easier to just set `mPinnedInodes = Preconditions.checkNotNull(pinnedInodes)` and something similar for locked blocks?

calvinjia commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,126 @@
+   * @return committed bytes for this dir
+   */
+  public long getCommittedBytes() {
+    // TODO: does pinedList, etc. change this value?
```
Yes, this should actually be something like `getEvictableBytes`

calvinjia commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,126 @@
+   * @param initialBlockSize
+   * @return a new TempBlockMeta under the underlying directory.
+   */
+  public TempBlockMeta CreateTempBlockMeta(long userId, long blockId, long initialBlockSize) {
```
Why does this method have a leading capital letter?

ooq replied:
Will fix the capital letter. It will be used by evictors/allocators to hide StorageDir from them.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+   */
+  public boolean isBlockLocked(long blockId) {
+    return mLockedBlocks.contains(blockId);
+  }
```
Is there a point in making a distinction between locked and pinned? Maybe it should just be `isBlockProtected` or something like that?

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+  public List<StorageTierView> getTierViewsBelow(int tierAlias) throws IOException {
+    // TODO: similar concern as in getTierView
+    int level = getTierView(tierAlias).getTierViewLevel();
+    return mTierViews.subList(level + 1, mTierViews.size());
```
This could throw index out of bounds if the lowest tier is passed in, should we handle that and return empty list?

ooq replied:
Do you mean from `sublist`? I think `a.sublist(a.size(), a.size())` returns a empty list.

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+  /**
+   * Redirecting to {@link BlockMetadataManager#getAvailableBytes(BlockStoreLocation)}
+   *
+   * @param location location the check available bytes
```
`the` -> `to`? Also maybe specify its a `BlockStoreLocation`

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+   */
+  public BlockMeta getBlockMeta(long blockId) throws IOException {
+    if (mPinnedInodes.contains(BlockInfo.computeInodeId(blockId))
+        || mLockedBlocks.contains(blockId)) {
```
Could we use the methods in this class for this logic?

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java:
```diff
@@ -0,0 +1,142 @@
+import tachyon.worker.block.meta.StorageTierView;
+
+/**
+ * This class exposes a narrower view of {@link BlockMetadataManager} to Evictors and Allocators.
```
Could you be a bit more specific about `narrower`?

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockWorker.java:
```diff
@@ -191,6 +196,10 @@ public void process() {
     mSyncExecutorService.submit(mBlockMasterSync);
+
+    // Start the pinlist syncer to perform the periodical fetching
+    mSyncExecutorService.submit(mPinListSync);
```
Should the threadpool size of the executor be increased?

calvinjia commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Task that fetches pinlist from master periodically
+ *
```
Extra line here

calvinjia commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,127 @@
+  /** Milliseconds between each fetch */
+  private final int mFetchIntervalMs;
+  /** Milliseconds between fetches before a timeout */
+  private final int mFetchTimeoutMs;
```
Maybe rename `fetch` to `sync` (or vice versa)? Otherwise there are too many similar terms being used.

calvinjia commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,127 @@
+  private volatile boolean mRunning;
+
+  // constructor for PinListSync, fall back to fetch paramters
+  PinListSync(BlockDataManager blockDataManager, TachyonConf tachyonConf) {
```
Is this visibility intended to be default?

calvinjia commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -74,6 +74,8 @@
   private final Evictor mEvictor;
   private List<BlockStoreEventListener> mBlockStoreEventListeners =
       new ArrayList<BlockStoreEventListener>();
+  /** A set of pinned blocks fetched from the master */
```
They aren't really blocks, more like block prefixes?

calvinjia commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -89,6 +91,10 @@ public TieredBlockStore(TachyonConf tachyonConf) throws IOException {
 
+    // initially use empty lists to provide full view
+    BlockMetadataManagerView initView = new BlockMetadataManagerView(mMetaManager,
+        new HashSet<Integer>(), new HashSet<Long>());
```
Is this planned to be used later?

ooq replied:
Yes, but I'll remove them for now and add these lines when I update the evictors and allocators.

calvinjia commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -477,4 +490,15 @@ private void freeSpaceInternal(long userId, long availableBytes, BlockStoreLocat
+   * @param blocks a list of IDs of block that have been pinned
+   */
+  @Override
+  public void setPinnedInodes(Set<Integer> inodes) {
```
Maybe make this thread safe?

calvinjia commented on servers/src/main/java/tachyon/worker/block/TieredBlockStore.java:
```diff
@@ -413,6 +419,13 @@ private void removeBlockNoLock(long userId, long blockId) throws IOException {
+    // TODO: update the view object instead of creating new one every time
+    return new BlockMetadataManagerView(mMetaManager, mPinnedInodes,
+        mLockManager.getLockedBlocks());
+  }
```
Do you need to consider concurrency with `setPinnedInodes`? Since its possible to get an empty set after the clear?

calvinjia commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,126 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jetty.server.UserIdentity;
```
What's this for?

calvinjia commented on servers/src/main/java/tachyon/worker/block/BlockLockManager.java:
```diff
@@ -200,6 +201,19 @@ public void cleanupUser(long userId) {
+   * @return a set of locked blocks
+   */
+  public Set<Long> getLockedBlocks() {
+    Set<Long> set = new HashSet<Long>();
```
Could you rename this to something more descriptive?

ooq replied:
For example?

calvinjia commented:
@ooq Thanks for the patch, I've left some comments.

ooq commented:
@apc999 @calvinjia I made another pass to the javadoc of all new files, addressed the comments and added two new methods. `BlockMetadataManagerView#isBlockEvictable` and `StorageDirView#getEvictableBytes`. Please take a look. Thanks!

apc999 commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,134 @@
+   * @param tachyonConf
+   * @return PinListSync constructed
+   */
+  PinListSync(BlockDataManager blockDataManager, TachyonConf tachyonConf) {
```
`public`?

apc999 commented on servers/src/main/java/tachyon/worker/block/PinListSync.java:
```diff
@@ -0,0 +1,134 @@
+  /**
+   * Constructor for PinListSync
+   *
+   * @param blockDataManager
```
describe parameters

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,146 @@
+   *
+   * @return underlying directory
+   */
+  public StorageDir getDirForCreatingBlock() {
```
Remove this method

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,146 @@
+  /**
+   * Create a TempBlockMeta given userId, blockId, and initialBlockSize.
+   *
+   * @param userId
```
In javadoc, put description of parameters.

apc999 commented:
+1 LGTM

calvinjia commented:
+1 LGTM
